runtime/lua/vim/treesitter/highlighter.lua

```diff
@@ -265,6 +265,59 @@ function TSHighlighter._on_win(_, _win, buf, _topline)
   return true
 end
 
+--- Highlights a region of the buffer with a given language
+---@param buf buffer buffer to highlight. Defaults to the current buffer if 0
+---@param ns number namespace for the highlights
+---@param region {[1]:number, [2]:number, [3]: number, [4]: number} (table) Region to highlight {start_row, start_col, end_row, end_col}
+---@param lang string treesitter language
+function TSHighlighter.highlight_region(buf, ns, region, lang)
+  buf = (buf == 0 or buf == nil) and vim.api.nvim_get_current_buf() or buf
+  vim.fn.bufload(buf)
+
+  vim.api.nvim_buf_clear_namespace(buf, ns, region[1], region[3] + 1)
+
+  local language = require('vim.treesitter.language')
+  language.require_language(lang)
+
+  -- we can't use a cached parser here since that could interfer with the existing parser of the buffer
+  local LanguageTree = require('vim.treesitter.languagetree')
+  local parser = LanguageTree.new(buf, lang)
+  parser:set_included_regions({ { region } })
+  parser:parse()
+
+  parser:for_each_tree(function(tstree, tree)
+    if not tstree then
+      return
+    end
+
+    local highlighter_query = TSHighlighterQuery.new(tree:lang())
+
+    -- Some injected languages may not have highlight queries.
+    if not highlighter_query:query() then
+      return
+    end
+
+    local iter = highlighter_query:query():iter_captures(tstree:root(), buf, region[1], region[3])
+
+    for capture, node, metadata in iter do
+      local start_row, start_col, end_row, end_col = node:range()
+
+      local hl = highlighter_query.hl_cache[capture]
+      local is_spell = highlighter_query:query().captures[capture] == 'spell'
+
+      if hl and not is_spell then
+        vim.api.nvim_buf_set_extmark(buf, ns, start_row, start_col, {
+          end_line = end_row,
+          end_col = end_col,
+          hl_group = hl,
+          priority = (tonumber(metadata.priority) or 100) + 10, -- add 10, so it will be higher than the standard highlighter of the buffer
+          conceal = metadata.conceal,
+        })
+      end
+    end
+  end, true)
```

Comment:
This is all duplicated from the highlighters `on_line` callback.

Can you DRY this up?

Reply:
I thought about that, but while it is similar, it also has a lot of differences.